src/main/scala/cognite/spark/v1/AssetsHierarchyBuilder.scala

```diff
@@ -0,0 +1,223 @@
+    } else {
+      IO.unit
+    }
+
+  def getCdfRootAndTree(sourceRootExternalId: CogniteExternalId): IO[(Option[Asset], List[Asset])] = {
+    val cdfTree = client.assets
+      .filter(AssetsFilter(rootIds = Some(Seq(sourceRootExternalId))))
+      .compile
+      .toList
+
+    for {
+      tree <- cdfTree
+      root <- IO(tree.find(_.externalId.contains(sourceRootExternalId.externalId)))
+    } yield (root, tree)
+  }
+
+  def upsertRoot(
+      root: AssetsIngestSchema,
+      sourceRoot: Option[Asset]
+  ): IO[Unit] =
+    sourceRoot match {
+      case None =>
+        client.assets.create(
+          Seq(
+            AssetCreate(
+              root.name,
+              None,
+              root.description,
+              root.source,
+              Some(root.externalId),
+              root.metadata,
+              None
+            ))) *> IO.unit
+      case Some(asset) =>
+        if (isMostlyEqual(root, asset)) {
+          IO.unit
+        } else {
+          client.assets.updateByExternalId(
+            Seq(
+              root.externalId -> AssetsIngestSchema
+                .toAssetUpdate(root)
+                .copy(parentExternalId = None)).toMap) *> IO.unit
+        }
+    }
+
+  def nodesToDeleteInsertUpdate(
+      root: AssetsIngestSchema,
+      leaves: Set[AssetsIngestSchema],
+      cdfTree: Seq[Asset]): (Seq[Asset], Set[AssetsIngestSchema], Set[AssetsIngestSchema]) = {
+    // Insert assets that are not present in CDF
+    val toInsert = leaves.filter(a => !cdfTree.map(_.externalId.get).contains(a.externalId))
+
+    // Keep assets that are in both source and cdf, delete those only in CDF
+    val (toKeep, toDelete) =
+      cdfTree.partition(a => (Seq(root) ++ leaves).map(_.externalId).contains(a.externalId.get))
+
+    // Filter out assets that would not be changed in an update
+    val toUpdate = leaves.filter(a => needsUpdate(a, toKeep))
+
+    (toDelete, toInsert, toUpdate)
+  }
+
+  def needsUpdate(leaf: AssetsIngestSchema, cdfTree: Seq[Asset]): Boolean = {
+    // Find the matching Asset in CDF
+    val cdfAsset = cdfTree.find(a => a.externalId.contains(leaf.externalId))
+    cdfAsset match {
+      case Some(asset) =>
+        // Find the parent Asset in CDF to be able to get the externalId of the parent
+        cdfTree.find(a => asset.parentId.contains(a.id)) match {
+          case Some(parentAsset) =>
+            !isMostlyEqual(leaf, asset) ||
+              !parentAsset.externalId.contains(leaf.parentExternalId)
+          case None =>
+            false
+        }
+      case None => false
+    }
+  }
+
+  def isMostlyEqual(leaf: AssetsIngestSchema, asset: Asset): Boolean =
+    leaf.description == asset.description &&
+      leaf.metadata == asset.metadata &&
+      leaf.name == asset.name &&
+      leaf.source == asset.source
+
+  def validateAndOrderInput(
+      tree: Array[AssetsIngestSchema]): (AssetsIngestSchema, Set[AssetsIngestSchema]) = {
```

Comment:
This is likely my fault after bad comments on a previous review :( but there's no order to a `Set` so it should probably be a `Seq` (or `List`, or whatever).
Our tests should catch that though, right?

Reply:
Tests don't catch it yet, I suppose because the asset trees are small enough to always fit in a single batch. We need to ensure that the batching happens in the right places though, or this won't work for asset trees with more than 1,000 assets.